Add new x-pack endpoints to track the progress of a search asynchronously

jimczi opened the pull request:
### High level view

This change introduces a new API in x-pack basic that allows to track the progress of a search.
Users can submit an asynchronous search through a new endpoint called `_async_search` that
works exactly the same as the `_search` endpoint but instead of blocking and returning the final response when available, it returns a response after a provided `wait_for_completion` time.

````
# Submit an _async_search and waits up to 100ms for a final response
GET my_index_pattern*/_async_search?wait_for_completion=100ms
{
  "aggs": {
    "date_histogram": {
      "field": "@timestamp",
      "fixed_interval": "1h"
    }
  }
}
````

If after 100ms the final response is not available, a `partial_response` is included in the body:

````
{
  "id": "9N3J1m4BgyzUDzqgC15b",
  "version": 1,
  "is_running": true,
  "is_partial": true,
  "response": {
   "_shards": {
       "total": 100,
       "successful": 5,
       "failed": 0
    },
    "total_hits": {
      "value": 1653433,
      "relation": "eq"
    },
    "aggs": {
      ...
    }
  }
}
````

The partial response contains the total number of requested shards, the number of shards that successfully returned and the number of shards that failed.
It also contains the total hits as well as partial aggregations computed from the successful shards.
To continue to monitor the progress of the search users can call the get `_async_search` API like the following:

````
GET _async_search/9N3J1m4BgyzUDzqgC15b/?wait_for_completion=100ms
````

That returns a new response that can contain the same partial response than the previous call if the search didn't progress, in such case the returned `version`
should be the same. If new partial results are available, the version is incremented and the `partial_response` contains the updated progress.
Finally if the response is fully available while or after waiting for completion, the `partial_response` is replaced by a `response` section that contains the usual _search response:

````
{
  "id": "9N3J1m4BgyzUDzqgC15b",
  "version": 10,
  "is_running": false,
  "response": {
     "is_partial": false,
     ...
  }
}
````

## Persistency

Asynchronous search are stored in a restricted index called `.async-search` if they survive (still running) after the initial submit. Each request has a keep alive that defaults to 5 days but this value can be changed/updated any time:
`````
GET my_index_pattern*/_async_search?wait_for_completion=100ms&keep_alive=10d
`````
The default can be changed when submitting the search, the example above raises the default value for the search to `10d`. 
`````
GET _async_search/9N3J1m4BgyzUDzqgC15b/?wait_for_completion=100ms&keep_alive=10d
`````
The time to live for a specific search can be extended when getting the progress/result. In the example above we extend the keep alive to 10 more days.
A background service that runs only on the node that holds the first primary shard of the `async-search` index is responsible for deleting the expired results. It runs every hour but the expiration is also checked by running queries (if they take longer than the keep_alive) and when getting a result.

Like a normal `_search`, if the http channel that is used to submit a request is closed before getting a response, the search is automatically cancelled. Note that this behavior is only for the submit API, subsequent GET requests will not cancel if they are closed. 

## Resiliency

Asynchronous search are not persistent, if the coordinator node crashes or is restarted during the search, the asynchronous search will stop. To know if the search is still running or not the response contains a field called `is_running` that indicates if the task is up or not. It is the responsibility of the user to resume an asynchronous search that didn't reach a final response by re-submitting the query. However final responses and failures are persisted in a system index that allows
to retrieve a response even if the task finishes.

````
DELETE _async_search/9N3J1m4BgyzUDzqgC15b
````

The response is also not stored if the initial submit action returns a final response. This allows to not add any overhead to queries that completes within the initial `wait_for_completion`.

## Security

The `.async-search` index is a restricted index (should be migrated to a system index in +8.0) that is accessible only through the async search APIs. These APIs also ensure that only the user that submitted the initial query can retrieve or delete the running search. Note that admins/superusers would still be able to cancel the search task through the task manager like any other tasks.

Relates #49091

jimczi committed:
Add new x-pack endpoints to asynchronously track the progress of a search

elasticmachine commented:
Pinging @elastic/es-search (:Search/Search)

jimczi commented:
@elasticmachine run elasticsearch-ci/packaging-sample-matrix

jimczi committed:
notify partial reduce even on top-docs query

jimczi committed:
fix npe

jpountz reviewed (commented):
Woohoo! This looks great in general, some questions:
 - If I read correctly, it's up to the user to garbage collect responses manually. Should we do this automatically when a final response has been retrieved? We already have a `wait_for_completion` parameter that allows to reduce the number of roundtrips for fast requests, so it doesn't feel consistent to always require a new roundtrip to delete the response? I'm also a bit biased towards reducing the number of cases when responses need to be garbage-collected via ILM, as you could accumulate a large volume of responses in 5 days?
 - If we moved from time-based ids to true uuids - which can't be guessed, I wonder whether we'd still need to require that the user that views a response is the same as the user who submitted the request. I don't think it would be surprising to users that sharing the id of an async search has pretty much the same consequences as sharing the response of the search request?
 - Since `response` and `partial_response` should have mostly the same format, I wonder whether we should use the same `response` key combined with a `partial` flag?

jimczi commented:
> If I read correctly, it's up to the user to garbage collect responses manually. Should we do this automatically when a final response has been retrieved? We already have a wait_for_completion parameter that allows to reduce the number of roundtrips for fast requests, so it doesn't feel consistent to always require a new roundtrip to delete the response? I'm also a bit biased towards reducing the number of cases when responses need to be garbage-collected via ILM, as you could accumulate a large volume of responses in 5 days?

I like the idea, getting the same final response twice is something that our regular caches should handle transparently so this would also emphasize the fact that this response are not meant to be used as an additional cache.

> If we moved from time-based ids to true uuids - which can't be guessed, I wonder whether we'd still need to require that the user that views a response is the same as the user who submitted the request. I don't think it would be surprising to users that sharing the id of an async search has pretty much the same consequences as sharing the response of the search request?

+1 for true uuids, I agree with the response sharing analogy but since we want to delete final responses when they are reported back I think it would be nice to have this extra layer. It's not a lot of work and something that we already implement in scrolls.

> Since response and partial_response should have mostly the same format, I wonder whether we should use the same response key combined with a partial flag?

Are you talking of the rest format or the internal response ? I think it's important to keep the distinction internally (for the hlrc) but I agree that the response could look like this:
````
{
  "id": "9N3J1m4BgyzUDzqgC15b",
  "version": 1,
  "is_running": true,
  "response": {
      "is_partial": true
````
Is it what you meant ?

jpountz commented:
Yes that's what I meant.

giladgal commented:
This all sounds great. Two suggestions to consider:
1. If there is a burst of queries for which the result sets are not fully retrieved (e.g. because they are lengthy and the user checks them on a long interval). Can we crash the system because there are too many result sets? Should we start deleting based on the index size of the results index (not just on time or retrieval)?

2. Would it make sense to have a limit on the time that the async query runs, e.g. have a default for the environment and allow users to specifically indicate that their query should be killed after a certain time interval (e.g. kill this query if it doesn't conclude after 24 hours). My concern is that if we don't have a limit, we can end up having zombie queries that queue on a resource starved system because there is no time-out to kill them. Such a mechanism would be an automated way for the admin to kill queries based on a very lengthy timeout.

jimczi committed:
use a single object named response in the xcontent serialization of the async search response

jimczi committed:
delete frozen search response automatically

jimczi committed:
add integration tests

jimczi committed:
simplify schema for the .async-search index

jimczi committed:
switch to true random uuuids

jimczi commented:
I pushed another iteration that addresses @jpountz's comments. 
We now use true random uuids for the async search index and the document is automatically deleted when the final response (completed or failed) is returned to the user (through submit or get). 

We also discussed the best options to secure the system index with @elastic/es-security.  The simplest way today would be to add the async search index in the `RestrictedIndicesNames` like we do  for the ` .security`  index. Only the superuser would be able to access this index directly which is enough imo. The restricted indices don't support index pattern today but @albertzaharovits said that it would  be trivial  to add. The other option would  be to use a single index and to replace the ilm garbage collection with a periodic delete by query. I am leaning towards the first option (adding support for index patterns in `RestrictedIndicesNames`)  since it simpler and ensures that we re-create the index periodically but I am curious to hear what others think.
I'll now focus on securing the async-search APIs to ensure that only the user that submitted the query can retrieve or delete the running search. Note that admins/superusers would still be able to cancel the search task through the task manager like any other tasks.

jimczi committed:
iter

colings86 commented:
I wonder if we should add "pending_shards" to the partial_response section so it's explicit that we are expecting those shard to return? 

Also I wonder if we should group the shard counts together into its own object?

jimczi committed:
replace waitUntil with assertBusy

mark-vieira commented:
FYI, you'll need to merge in the latest changes from `master` to fix these CI failure due to the new Java 13 build requirement.

jimczi committed:
add a security layer that restricts the usage of the async search APIs to the user that submitted the initial rquest

jimczi committed:
Merge branch 'master' into async_search

jimczi committed:
fix checkstyle

jimczi committed:
fix double clean

jimczi committed:
iter

itsmed commented:
aborted build for the upgrade, jenkins test this please

jimczi committed:
Merge branch 'master' into async_search

jimczi committed:
add more logging

jimczi commented:
run elasticsearch-ci/1

jimczi commented:
run elasticsearch-ci/2

albertzaharovits commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchStoreService.java:
```diff
@@ -0,0 +1,165 @@
+     */
+    void getResponse(AsyncSearchId searchId, ActionListener<AsyncSearchResponse> listener) {
+        final Authentication current = Authentication.getAuthentication(client.threadPool().getThreadContext());
+        GetRequest internalGet = new GetRequest(searchId.getIndexName())
```
Because the result index is encoded in the asyncTaskId, a user can create his own index, and then craft an asyncTaskId that the target user can use to get a bogus result.

jimczi replied:
Is is an issue ? The index name is resolved in the submit action so changing the index when retrieving an id should not impact the storage of the "real" response. In what scenario do you see this happening ? The worst case is that we return the "fake" response of the other index but we'll never return the "real" response. We can also add a validation that the index name matches the async-search alias since it's a protected namespace. Would that solve your concern ?

albertzaharovits replied:
It is not a big problem in the same sense that users can alter one another 's responses, but being able to craft responses is still problematic.

I prefer we offer users the following guarantee: If you call the get response API you get the response to a job that you submitted, OR an error response (missing) if the job is expired or otherwise missing (node restarted) or if the job-id is not yours. Somebody else should not be able to craft responses that you interpret as valid.

Does it make sense?

jimczi replied:
Sure, that makes sense. I can add a validation that checks if the index name matches the async search index pattern. That would be enough imo, do you agree ?

albertzaharovits replied:
Yes, I think that would work.

jimczi replied:
I pushed https://github.com/elastic/elasticsearch/pull/49931/commits/baa498d84d71d87c4afad294845c3c4f68853672 to validate the index name and to fail the request if the index name does not match the index pattern.

jimczi replied:
I pushed the [validation](https://github.com/elastic/elasticsearch/pull/49931/commits/d03b9402192b03a592e08907e07b9ac10f5f5515) in the decoding of the search id, this means that it will be checked on all APIs.

albertzaharovits commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportGetAsyncSearchAction.java:
```diff
@@ -0,0 +1,235 @@
+            return false;
+        }
+        Authentication origin = Authentication.decode(originHeaders.get(AUTHENTICATION_KEY));
+        return ensureAuthenticatedUserIsSame(origin, current);
```
I think this works well, but it probably needs more tests.

jimczi committed:
replace routing with preference in get requests

jimczi committed:
iter

This commit ensures that the rollover index is created before launching an async search.
It also changes the design to not store any document if the submit action returns a final response.

jimczi committed:
validate the index name when receiving a search id

jimczi commented:
I pushed another iteration that solves the first level of security.  The authenticated user that initiated the request is now the only one that can access responses and progress. For the second level, @albertzaharovits is now adding the support to protect the `.async-search` namespace from any external access: https://github.com/elastic/elasticsearch/pull/50294.
I also talked to @henningandersen and he suggested to delay the storage of the response to after the initial wait for completion. The idea is that we don't need to store the response (or any placeholder) if the initial request (submit) already returned the final response. In such case we just skip storing a response. This is implemented in https://github.com/elastic/elasticsearch/pull/49931/commits/177e3039f78258ae8bec81e76a1ef7699f81fa4a and I changed the description to explain this behavior.
I removed the WIP label since it is quite advanced now and will work on the documentation.

jimczi committed:
Merge branch 'master' into async_search

jimczi committed:
add authentication test

jimczi committed:
Merge branch 'master' into async_search

jimczi committed:
fix block plugin to throw exceptions in createWeight rather than toQuery

jimczi commented:
I discussed with @lukasolson  for the integration in Kibana and one thing I wonder, now that `submit` can return a final response without storing anything, is whether we need to clean automatically on completion for the subsequent `get` requests. It seems natural to me to leave the responsibility to delete the response to users in the case that the `submit` request didn't complete within the provided `wait_for_completion`. Fast requests won't have any overhead so only long running queries (that survived the initial submit) would need a deletion and if the user forgets or fails to delete then the ILM policy will delete the response after a certain period. WDYT ?

jimczi committed:
remove clean_on_completion option after the initial submit

jimczi committed:
Merge branch 'master' into async_search

jimczi committed:
move shard stats in a _shards section for partial response

jimczi commented:
> Also I wonder if we should group the shard counts together into its own object?

I pushed a [change](https://github.com/elastic/elasticsearch/pull/49931/commits/93189d38054961070647eaf9734c922af275c720) to move the counts in a `_shards` object like the normal search response.

albertzaharovits commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportGetAsyncSearchAction.java:
```diff
@@ -0,0 +1,132 @@
+    @Override
+    protected void doExecute(Task task, GetAsyncSearchAction.Request request, ActionListener<AsyncSearchResponse> listener) {
+        try {
+            AsyncSearchId searchId = AsyncSearchId.decode(request.getId());
```
Just a nit, but failure to decode will translate to a 500 , but it should be a 400.
I think it works if you throw `IllegalArgumentException` when decoding fails.

jimczi replied:
++, I pushed https://github.com/elastic/elasticsearch/pull/49931/commits/d03b9402192b03a592e08907e07b9ac10f5f5515

albertzaharovits reviewed (commented):
I did a thorough review, but I've not yet looked closely over the tests.
I'll look into the tests as the last step before approving.

jimczi commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchStoreService.java:
```diff
@@ -0,0 +1,288 @@
+     */
+    static AsyncSearchResponse decodeResponse(String value, NamedWriteableRegistry registry) throws IOException {
+        try (ByteBufferStreamInput buf = new ByteBufferStreamInput(ByteBuffer.wrap(Base64.getDecoder().decode(value)))) {
+            try (StreamInput in = new NamedWriteableAwareStreamInput(buf, registry)) {
```
They are needed to read aggregations which use named writeable.

jimczi committed:
address review

jimczi commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchId.java:
```diff
@@ -0,0 +1,105 @@
+            out.writeString(indexName);
+            out.writeString(docId);
+            out.writeString(taskId.toString());
+            return Base64.getEncoder().encodeToString(BytesReference.toBytes(out.bytes()));
```
++, I pushed https://github.com/elastic/elasticsearch/pull/49931/commits/d03b9402192b03a592e08907e07b9ac10f5f5515

jimczi commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportSubmitAsyncSearchAction.java:
```diff
@@ -0,0 +1,196 @@
+            return;
+        }
+        final String docID = UUIDs.randomBase64UUID(random);
+        store.ensureAsyncSearchIndex(clusterService.state(), new ActionListener<>() {
```
++, I pushed https://github.com/elastic/elasticsearch/pull/49931/commits/d03b9402192b03a592e08907e07b9ac10f5f5515

jimczi commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchStoreService.java:
```diff
@@ -0,0 +1,288 @@
+    AsyncSearchStoreService(TaskManager taskManager, ThreadPool threadPool, Client client, NamedWriteableRegistry registry) {
+        this.taskManager = taskManager;
+        this.threadPool = threadPool;
+        this.client = new OriginSettingClient(client, INDEX_LIFECYCLE_ORIGIN);
```
We'll add a new origin so this is just temporary ?

albertzaharovits replied:
I think a new origin is a nice-to-have . I would say that a "tasks" origin for async search tasks is good enough? What do you think?

jimczi replied:
We discussed offline and agreed that the "tasks" origin is good enough for now so I pushed https://github.com/elastic/elasticsearch/commit/7ffb1dde347870e4541ad05532b1acdeaccf2b14.

jimczi commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportDeleteAsyncSearchAction.java:
```diff
@@ -0,0 +1,60 @@
+        try {
+            AsyncSearchId searchId = AsyncSearchId.decode(request.getId());
+            // check if the response can be retrieved by the user (handle security) and then cancel/delete.
+            store.getResponse(searchId, ActionListener.wrap(res -> cancelTaskAndDeleteResult(searchId, listener), listener::onFailure));
```
The searchId is now checked on decoding (https://github.com/elastic/elasticsearch/pull/49931/commits/d03b9402192b03a592e08907e07b9ac10f5f5515) so we should be protected against this now.

jimczi commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportDeleteAsyncSearchAction.java:
```diff
@@ -0,0 +1,60 @@
+
+    private void cancelTaskAndDeleteResult(AsyncSearchId searchId, ActionListener<AcknowledgedResponse> listener) {
+        try {
+            nodeClient.execute(CancelTasksAction.INSTANCE, new CancelTasksRequest().setTaskId(searchId.getTaskId()),
```
I switched to the `client` with origin but the idea is still to use a new origin for the async search.

albertzaharovits replied:
Okay, we'll be adding a new origin.

jimczi commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportSubmitAsyncSearchAction.java:
```diff
@@ -0,0 +1,196 @@
+    private void onTaskCompletion(ThreadPool threadPool, AsyncSearchTask task,
+                                    AtomicReference<Boolean> reference) {
+        final Boolean shouldStoreResult = reference.get();
+        if (shouldStoreResult == null) {
```
I don't understand, we schedule a retry so that we don't need to block the current thread. 
What protection should be added ?

jimczi replied:
We discussed offline and decided to keep the schedule of the retry through the thread pool in order to not block a thread with a notify/wait.

jpountz replied:
Would it be possible to avoid polling still? E.g. by asking `shouldStoreResult` to notify us when its value is set.

jimczi commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchId.java:
```diff
@@ -0,0 +1,105 @@
+     */
+    static String encode(String indexName, String docId, TaskId taskId) {
+        try (BytesStreamOutput out = new BytesStreamOutput()) {
+            out.writeString(indexName);
```
Why would it be needed ? We don't have versioned id and I don't think we'll need it ?

albertzaharovits replied:
My thinking was that a version could be useful in a mixed cluster scenario. But on further consideration, I now think it's not necessary.

jimczi commented on x-pack/plugin/core/src/main/resources/async-search-history.json:
```diff
@@ -0,0 +1,29 @@
+      }
+    }
+  }
+}
```
The pr now uses a simple mapping with disabled objects and we extract everything from base64 string in the _source directly.

jimczi committed:
Merge branch 'master' into async_search

jimczi committed:
address review

jimczi commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:
```diff
@@ -206,6 +206,9 @@ private static boolean shouldAuthorizeIndexActionNameOnly(String action, Transpo
             case MultiSearchAction.NAME:
+            case "indices:data/read/async_search/submit":
+            case "indices:data/read/async_search/get":
+            case "indices:data/read/async_search/delete":
```
+1, I pushed https://github.com/elastic/elasticsearch/pull/49931/commits/7ffb1dde347870e4541ad05532b1acdeaccf2b14 to address your comment.

jimczi committed:
unused import

jimczi committed:
Merge branch 'master' into async_search

jimczi committed:
line len

albertzaharovits reviewed (approved):
LGTM. I've left a few minor suggestions for tests, but addressing those should not prevent merging.

jpountz commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchTask.java:
```diff
@@ -0,0 +1,177 @@
+         * Ensure that we're performing the final reduce only when users explicitly requested
+         * a response through a {@link GetAsyncSearchAction.Request}.
+         */
+        public synchronized AsyncSearchResponse get(boolean doFinalReduce) {
```
why do we need the `synchronized` keyword?

jimczi replied:
I wanted to avoid performing the final reduce multiple times but I simplified the logic and removed the synchronization in https://github.com/elastic/elasticsearch/pull/49931/commits/9056b873039b5f087a7f98dee1fd31d3850dde11 since we expect a single user to access the response.

jpountz commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportGetAsyncSearchAction.java:
```diff
@@ -0,0 +1,126 @@
+                        new ActionListenerResponseHandler<>(listener, AsyncSearchResponse::new, ThreadPool.Names.SAME));
+                }
+            }
+        } catch (IOException e) {
```
should we catch all exceptions?

jimczi replied:
It shouldn't be necessary but I pushed https://github.com/elastic/elasticsearch/pull/49931/commits/9056b873039b5f087a7f98dee1fd31d3850dde11 to be on the safe side ;).

jpountz commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportSubmitAsyncSearchAction.java:
```diff
@@ -0,0 +1,189 @@
+        this.reduceContextSupplier = () -> searchService.createReduceContext(true);
+        this.searchAction = searchAction;
+        this.store = new AsyncSearchStoreService(taskManager, threadPool, client, registry);
+        this.random = new Random(System.nanoTime());
```
This looks dangerous to me since it would be legal for System#nanoTime to have a resolution of tens of milliseconds, so two requests could get the same seed. I wonder why you don't call `UUIDs.randomBase64UUID()` (without a Random parameter) below, which uses a shared secure random?

jimczi replied:
+1 thanks, I moved to `UUIDs.randomBase64UUID()` in https://github.com/elastic/elasticsearch/commit/9056b873039b5f087a7f98dee1fd31d3850dde11

jpountz commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportSubmitAsyncSearchAction.java:
```diff
@@ -0,0 +1,189 @@
+                                    shouldStoreResult.set(true);
+                                    submitListener.onResponse(response);
+                                }));
+                        } catch (IOException exc) {
```
should we catch all exceptions?

jimczi replied:
+1, I pushed https://github.com/elastic/elasticsearch/commit/9056b873039b5f087a7f98dee1fd31d3850dde11

jpountz commented on x-pack/plugin/core/src/main/resources/async-search.json:
```diff
@@ -0,0 +1,29 @@
+  },
+  "mappings": {
+    "_doc": {
+      "dynamic": false,
```
let's make it strict?

jimczi replied:
Sure, I pushed https://github.com/elastic/elasticsearch/commit/9056b873039b5f087a7f98dee1fd31d3850dde11

jpountz reviewed (commented):
This looks good to me in general. I left some questions/comments.

jimczi committed:
Merge branch 'master' into async_search

jimczi committed:
address review

jimczi committed:
add more tests

jimczi commented on server/src/main/java/org/elasticsearch/action/search/SearchResponse.java:
```diff
@@ -108,6 +108,11 @@ public SearchResponse(SearchResponseSections internalResponse, String scrollId,
+        this(clone.internalResponse, clone.scrollId, clone.totalShards, clone.successfulShards, clone.skippedShards,
+            clone.tookInMillis, clone.shardFailures, clone.clusters);
+    }
+
```
I didn't want to overload the original class since it has its own definition of partial results (totalShards != successfulShards). The `partial` flag here is only relevant in the context of an async search.

jimczi commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchId.java:
```diff
@@ -0,0 +1,116 @@
+     */
+    static AsyncSearchId decode(String id) {
+        final AsyncSearchId searchId;
+        try (StreamInput in = new ByteBufferStreamInput(ByteBuffer.wrap( Base64.getUrlDecoder().decode(id)))) {
```
Thanks, I pushed https://github.com/elastic/elasticsearch/pull/49931/commits/9056b873039b5f087a7f98dee1fd31d3850dde11

jimczi commented on x-pack/plugin/async-search/qa/rest/src/test/resources/rest-api-spec/test/search/10_basic.yml:
```diff
@@ -0,0 +1,112 @@
+  - is_false: id
+  - match: { version:   7 }
+  - match: { response.is_partial: false }
+  - length: { response.hits.hits:        3 }
```
Thanks, I pushed https://github.com/elastic/elasticsearch/pull/49931/commits/9056b873039b5f087a7f98dee1fd31d3850dde11

jimczi commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchId.java:
```diff
@@ -0,0 +1,116 @@
+        final AsyncSearchId searchId;
+        try (StreamInput in = new ByteBufferStreamInput(ByteBuffer.wrap( Base64.getUrlDecoder().decode(id)))) {
+            searchId = new AsyncSearchId(in.readString(), in.readString(), new TaskId(in.readString()));
+        } catch (IOException e) {
```
+1, I pushed https://github.com/elastic/elasticsearch/pull/49931/commits/9056b873039b5f087a7f98dee1fd31d3850dde11

jimczi commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchStoreService.java:
```diff
@@ -0,0 +1,295 @@
+            // alias does not exist but initial index does, something is broken
+            andThen.onFailure(new IllegalStateException("async-search index [" + initialIndexName +
+                "] already exists but does not have alias [" + ASYNC_SEARCH_ALIAS + "]"));
+        } else if (current.isAlias() && current instanceof AliasOrIndex.Alias) {
```
+1, I pushed https://github.com/elastic/elasticsearch/pull/49931/commits/9056b873039b5f087a7f98dee1fd31d3850dde11

jimczi commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchStoreService.java:
```diff
@@ -0,0 +1,295 @@
+        source.put(HEADERS_FIELD, headers);
+        UpdateRequest request = new UpdateRequest().index(searchId.getIndexName()).id(searchId.getDocId())
+            .doc(source, XContentType.JSON)
+            .detectNoop(false);
```
we don't need it, it was added to avoid the noop detection since we know that the source has changed but I can remove it.

jimczi commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchStoreService.java:
```diff
@@ -0,0 +1,295 @@
+                    listener.onFailure(new ResourceNotFoundException("id [{}] not found", searchId.getEncoded()));
+                } else {
+                    listener.onResponse(new AcknowledgedResponse(true));
+                }
```
From looking at the code I don't think that's possible for delete request.

jimczi commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchTask.java:
```diff
@@ -0,0 +1,177 @@
+    }
+
+    private class Listener extends SearchProgressActionListener {
+        private int totalShards = -1;
```
onListShards (where the value is set) is mutually exclusive with the functions that access it so it's not required. However I changed the logic to use latches in https://github.com/elastic/elasticsearch/pull/49931/commits/9056b873039b5f087a7f98dee1fd31d3850dde11 in order to make the logic more readable. Can you take another look ?

jimczi commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchTask.java:
```diff
@@ -0,0 +1,177 @@
+    private class Listener extends SearchProgressActionListener {
+        private int totalShards = -1;
+        private AtomicInteger version = new AtomicInteger(0);
+        private AtomicInteger shardFailures = new AtomicInteger(0);
```
Done, in https://github.com/elastic/elasticsearch/pull/49931/commits/9056b873039b5f087a7f98dee1fd31d3850dde11

jimczi commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchTask.java:
```diff
@@ -0,0 +1,177 @@
+        private AtomicInteger shardFailures = new AtomicInteger(0);
+
+        private int lastSuccess = 0;
+        private int lastFailures = 0;
```
I simplified the logic in https://github.com/elastic/elasticsearch/pull/49931/commits/9056b873039b5f087a7f98dee1fd31d3850dde11, thanks

jimczi commented on x-pack/plugin/async-search/qa/rest/build.gradle:
```diff
@@ -0,0 +1,24 @@
+}
+
+check.dependsOn restTest
+test.enabled = false
```
It does, thanks for checking.

jimczi committed:
remove unrelated change

jimczi committed:
fix nocommit

jimczi committed:
add more statistics to the search progress listener (number of skipped shards, remote clusters statistics)

jimczi committed:
Replace generic thread pool execution with asynchronous listener that waits for completion in the task

jimczi committed:
Merge branch 'master' into async_search

jimczi committed:
fix checkstyle

jimczi commented:
I pushed another iteration that:
* Remove the usage of 206 and 304 in the http status code response. These codes are not suited for the usage that we intended so they are replaced by a simple 200/OK. Thanks to @DaveCTurner  for spotting this.
* I added more statistics in the search progress listener so that we can reuse the SearchResponse for partial content too.
* I changed the code to use asynchronous listener to wait for completion rather than spawning new threads in the generic threadpool.

@henningandersen  could you help review the asynchronous listener that I added in the AsyncSearchTask ? I implemented what we discussed offline so I am curious to hear what you think.

jimczi committed:
fix wrong assert

jimczi commented:
@elasticmachine run elasticsearch-ci/2

jimczi committed:
Expose the logic to cancel task when the rest channel is closed

This commit moves the logic that cancels search requests when the rest channel is closed
to a generic client that can be used by other APIs. This will be useful for any rest action
that wants to cancel the execution of a task if the underlying rest channel is closed by the
client before completion.

Relates #49931
Relates #50990
Relates #50990

henningandersen reviewed (commented):
I primarily reviewed `AsyncSearchTask`. I have a few smaller comments, but otherwise this is looking good.

jimczi committed:
address review

jimczi committed:
plug the automatic cancel of search if the rest channel is closed when submitting the initial async search request

jimczi commented:
Thanks @henningandersen , I pushed a commit that hopefully addresses all your comments:
https://github.com/elastic/elasticsearch/pull/49931/commits/6ae79030da84602b509e70f291687672c5a8b75a

jimczi commented:
I also cherry-picked the changes of https://github.com/elastic/elasticsearch/pull/51423 and plugged the new logic to ensure that we cancel the search if the user closes the rest channel on submit. This should validate that #51423 can be used effectively in another API than `_search`.
However the cancellation logic doesn't handle grand-children at the moment so we're waiting on https://github.com/elastic/elasticsearch/issues/50990 to fully address this issue.

jimczi committed:
Merge branch 'master' into async_search

henningandersen commented:
Thanks Jim, this looks great.

jimczi committed:
handle expiration time on a per request basis

jimczi committed:
use a single replica for the hidden index

jimczi committed:
Merge branch 'master' into async_search

jimczi committed:
Merge branch 'master' into async_search

jimczi committed:
iter

jimczi committed:
do not set id twice

jimczi committed:
change expectation in test now that we use one replica by default

jimczi committed:
Merge branch 'master' into async_search

jimczi committed:
fix x-pack user to allow restricted indicesé

jimczi committed:
cleanup

jimczi committed:
Allow x-pack user to manage async-search indices

The x-pack user has all privileges on all indices except security indices and restricted indices.
This change adds the authorization for restricted indices in order to allow the system user to manage
these indices internally. This will allow the async-search restricted indices to be accessed only
with the x-pack user.

jimczi committed:
iter

monfera commented:
Awesome PR @jimczi! 🎉 As this PR is rather close to being merged, is there a short description of what to expect, eg. if/once it's added to Kibana (or whichever client) eg. 
- does it give a ratio, ie. the user could estimate how much to still wait?
- what is it relative to, eg. duration to complete; CPU time to complete; workload eg. doc/agg count to chew through; something else?
- roughly, how reliable is the progress indication, eg. can it spend half the time at 99%?
- (if it's relevant) how does it relate to the query cancellation API?

I realize there are a ton of details above, followed by detailed discussion that might have altered some things, so it's more about the question of what to realistically expect on the receiving end

jimczi committed:
Merge branch 'master' into async_search

jimczi committed:
another big iter

jimczi commented:
> does it give a ratio, ie. the user could estimate how much to still wait?

It returns the number of shards that returned a result for the query. So the granularity is at the shard level, you can also check the partial `total_hits` that return the number of documents collected so far. 

> what is it relative to, eg. duration to complete; CPU time to complete; workload eg. doc/agg count to chew through; something else?

Only the number of shards vs the total. 

> roughly, how reliable is the progress indication, eg. can it spend half the time at 99%?

I guess so, we don't have a way currently to know the progress of a search in a single shard so depending on the activity on the cluster , size of shards, ... , one shard can take much longer to return a result than another.

> (if it's relevant) how does it relate to the query cancellation API?

Only for admins, users of this API have a dedicated delete API that cancels/remove the async search. We also automatically cancel the request if the http channel that is used to submit a search is closed before getting the first result (same as normal `_search`).

jimczi commented:
I updated the description of the pr with the current status and the things that are left to do. I pushed another big iteration yesterday, we now use a single index with one shard and one replica that is periodically checked to remove the expired results. This service is executed only by the node that hosts the primary shard of the `.async-search` index. I need to add more tests but I think it's ready for another round of review ;).

jimczi committed:
Merge branch 'master' into async_search

jimczi committed:
small fix after merging master

jimczi committed:
checkstyle

jimczi committed:
Merge branch 'master' into async_search

jimczi committed:
restore test

jimczi committed:
fix more test

jimczi committed:
iter

javanna reviewed (commented):
I am in the process of reviewing, not done yet, but I wanted to send out the comments I have so far.

javanna reviewed (commented):
A couple more comments from another review round I did. Nothing blocking. To me these are all follow-up changes that can be made once merged together with adding docs and the missing tests. The cancellation aspects are the most important to address I think.

jimczi committed:
Merge branch 'master' into async_search

jimczi committed:
address review and plug the new async_search origin

jimczi committed:
unused import

jimczi commented on server/src/main/java/org/elasticsearch/action/search/SearchResponse.java:
```diff
@@ -108,11 +108,20 @@ public SearchResponse(SearchResponseSections internalResponse, String scrollId,
+    public SearchResponseSections getInternalResponse() {
+        return internalResponse;
+    }
+
```
This is needed to access the `SearchResponseSections` all together. See ` MutableSearchResponse` for the usage of it.

javanna replied:
Yep, I see now how the getter is used. But I still see no usage of the new constructor. Is it me?

jimczi replied:
I removed the constructor already,  you were right it was not used :).

jimczi commented on server/src/main/java/org/elasticsearch/action/search/SearchShard.java:
```diff
@@ -34,7 +34,7 @@
     private final ShardId shardId;
 
-    SearchShard(@Nullable String clusterAlias, ShardId shardId) {
+    public SearchShard(@Nullable String clusterAlias, ShardId shardId) {
```
It is called from tests in this pr

javanna replied:
I see now how when I first left these comments my IDE was acting out and hadn't yet imported the newly added xpack code... so a lot of usages were just not loaded.

jimczi commented on server/src/test/java/org/elasticsearch/action/search/SearchProgressActionListenerIT.java:
```diff
@@ -189,7 +194,14 @@ public void onFailure(Exception e) {
+        client.executeLocally(SearchAction.INSTANCE, new SearchRequest(request) {
+            @Override
+            public Task createTask(long id, String type, String action, TaskId parentTaskId, Map<String, String> headers) {
+                SearchTask task = (SearchTask) super.createTask(id, type, action, parentTaskId, headers);
```
++, I pushed a change to do that

jimczi commented on test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/DoSection.java:
```diff
@@ -345,6 +345,7 @@ private String formatStatusCodeMessage(ClientYamlTestResponse restTestResponse,
     private static Map<String, Tuple<String, org.hamcrest.Matcher<Integer>>> catches = new HashMap<>();
 
     static {
+        catches.put("not_modified", tuple("304", equalTo(304)));
```
I reverted this change, this was  needed when the API was returning 304 but we're using this return code anymore.

jimczi commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchId.java:
```diff
@@ -0,0 +1,106 @@
+        try (BytesStreamOutput out = new BytesStreamOutput()) {
+            out.writeString(docId);
+            out.writeString(taskId.toString());
+            return Base64.getUrlEncoder().encodeToString(BytesReference.toBytes(out.bytes()));
```
The encoding of the id should not change and if it does we'll have to do it in bwc manner.  For the response, we store the version used to serialize the content so we'll be able to handle 2 major releases in the index.

jimczi commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportSubmitAsyncSearchAction.java:
```diff
@@ -0,0 +1,193 @@
+    private static final Logger logger = LogManager.getLogger(TransportSubmitAsyncSearchAction.class);
+
+    private final NodeClient nodeClient;
+    private final ThreadContext threadContext;
```
++, I removed it.

jimczi commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportSubmitAsyncSearchAction.java:
```diff
@@ -0,0 +1,193 @@
+        if (errors != null) {
+            submitListener.onFailure(errors);
+            return;
+        }
```
It was an oversight, I removed it, thanks.

jimczi commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/search/action/SubmitAsyncSearchRequest.java:
```diff
@@ -0,0 +1,169 @@
+        request.setCcsMinimizeRoundtrips(false);
+        request.setPreFilterShardSize(1);
+        request.setBatchedReduceSize(5);
+        request.requestCache(true);
```
I think it's ok to leave them open, some of them are already forbidden (CcsMinimizeRoundtrips and scroll) since they break the API but I don't see why we should prevent the other ones since this API should expose the same option than a normal search.

javanna replied:
agreed, my comment focused on stuff that if changed, may affect how the API works e.g. pre_filter_shard_size, batched_reduced_size. Can you show me where we reject ccs minimize roundtrips set to false?

jimczi commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/RestSubmitAsyncSearchAction.java:
```diff
@@ -0,0 +1,65 @@
+        ActionRequestValidationException validationException = submit.validate();
+        if (validationException != null) {
+            throw validationException;
+        }
```
it does but we do this in some rest layer to avoid creating a task if the validation fails. I can remove it if you think it's confusing

javanna replied:
I was not aware of it. I do find it confusing. If we need it here, we may need it in many other places too, then maybe validation should rather be done by some rest action base class?

jimczi replied:
I don't think it's needed either so I pushed a change to remove.

jimczi commented on server/src/main/java/org/elasticsearch/action/support/tasks/TransportTasksAction.java:
```diff
@@ -99,7 +99,12 @@ protected void doExecute(Task task, TasksRequest request, ActionListener<TasksRe
+        } catch (Exception exc) {
+            listener.onFailure(exc);
+            return;
+        }
```
That's a leftover from another attempt so I removed it. I think there's an issue here but I don't recall what it was ;). I'll try to tackle this in a different pr so I removed this change here.

jimczi commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchIndexService.java:
```diff
@@ -0,0 +1,339 @@
+        source.put(RESULT_FIELD, encodeResponse(response));
+        IndexRequest indexRequest = new IndexRequest(INDEX)
+            .id(docId)
+            .source(source, XContentType.JSON);
```
That would expose id collisions if it happens so I agree that we should add this option. I pushed a change to add it.

jimczi committed:
ensure that the task is unregistered before returning the final response to the user

jimczi commented:
Thanks for the review @javanna , I pushed some changes to address your comments and answered in the pr for those that don't require a change at the moment.

jimczi committed:
checkstyle

lukasolson commented:
It looks to me like this API is not properly handling the `rest_total_hits_as_int` parameter... I understand this is eventually going away but what is the short term plan for supporting it in this API?

jimczi commented:
> It looks to me like this API is not properly handling the rest_total_hits_as_int parameter... I understand this is eventually going away but what is the short term plan for supporting it in this API?

I wasn't planning to add the support for it since it's a new API and `rest_total_hits_as_int` is a BWC layer that was added to ease the upgrade from 6.x to 7.x. @lukasolson are you saying that this would be needed for Kibana in 7.x ?

lukasolson commented:
I think I can work around it for the time being.

javanna reviewed (approved):
I left a couple more small comments but LGTM. Can you also update the description and remove the mention of the 304 status code which I believe is outdated? Thanks for taking this to the finish line.

jimczi committed:
Merge branch 'master' into async_search

jimczi committed:
address review

jimczi committed:
unused import

jimczi committed:
Merge branch 'master' into async_search

jimczi commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/RestGetAsyncSearchAction.java:
```diff
@@ -0,0 +1,52 @@
+        ActionRequestValidationException validationException = get.validate();
+        if (validationException != null) {
+            throw validationException;
+        }
```
++

jimczi commented on x-pack/plugin/src/test/resources/rest-api-spec/api/async_search.get.json:
```diff
@@ -0,0 +1,38 @@
+      "keep_alive": {
+        "type": "time",
+        "description": "Specify the time that the request should remain reachable in the cluster."
+      }
```
Good catch, I added the support for this option in https://github.com/elastic/elasticsearch/pull/49931/commits/415c91034674506b915c732881ab6a784e5b9d55

jimczi committed:
fix rest API reference to the outdated 304 response status

javanna commented on x-pack/plugin/src/test/resources/rest-api-spec/api/async_search.submit.json:
```diff
@@ -0,0 +1,226 @@
+          "path":"/_async_search",
+          "methods":[
+            "GET",
+            "POST"
```
one more thing I had missed before, here we may want to remove GET? I tend to think that POST is the only method that suits an API that submits something. Was it here only for consistency with search?

javanna commented on x-pack/plugin/src/test/resources/rest-api-spec/api/async_search.get.json:
```diff
@@ -0,0 +1,42 @@
+        "type":"time",
+        "description":"Specify the time that the request should block waiting for the final response (default: 1s)"
+      },
+      "last_version":{
```
another small which we already discussed in another channel: I find naming not so clear for this parameter. If we do need it, I would love to discuss options for naming it differently as the version terminology may be confusing for users. I would love to find something more self explanatory regarding the behaviour and the use for this.

javanna commented on x-pack/plugin/src/test/resources/rest-api-spec/api/async_search.submit.json:
```diff
@@ -0,0 +1,226 @@
+      },
+      "keep_alive": {
+        "type": "time",
+        "description": "Specify the time that the request should remain reachable in the cluster."
```
maybe rephrase to something like "Specify the time interval in which the results (partial or final) for this search will be available"

jimczi committed:
remove last_version parameter

jimczi committed:
rephrase rest option after review

jimczi committed:
Merge branch 'master' into async_search

lizozom commented:
@jimczi 

Could you please provide an example on how to update the keep alive time of a currently running async search?

I tried this without success.

Also, the final GET returns all of the data on a single response, without pagination. Is that the way this should always be?

jimczi commented:
> Could you please provide an example on how to update the keep alive time of a currently running async search?

The response returns the initial keep alive instead of the updated one, I opened https://github.com/elastic/elasticsearch/pull/55435 to fix the bug.

> Also, the final GET returns all of the data on a single response, without pagination. Is that the way this should always be?

What do you mean by pagination ? It should return the same final response than a normal `_search`.
